src/frequenz/sdk/model/_manager.py

```diff
@@ -0,0 +1,366 @@
+
+        Args:
+            key: the key for which to build the model path.
+
+        Returns:
+            The model path.
+
+        Raises:
+            AssertionError: if the directory, file name, or the given key is
+                not set or allowed.
+        """
+        assert self.directory, "Directory must be set."
+        assert self.file_name, "File name must be set."
+        assert key in self._get_allowed_keys(), f"Key {key} is not allowed."
+        return f"{self.directory}/{key}/{self.file_name}"
+
+
+@dataclass
+class IntervalModelRepository(ModelRepository):
+    """Repository to manage interval-based models."""
+
+    directory: str
+    """The base directory where the interval-based models are stored."""
+
+    file_name: str
+    """The name of the file containing the model."""
+
+    num_models: int
+    """The number of interval-based models to load and manage."""
+
+    def _get_allowed_keys(self) -> list[Any]:
+        """Get the allowed keys for the interval-based models repository.
+
+        Returns:
+            the allowed keys for the interval-based models repository.
+        """
+        return list(range(self.num_models))
+
+    def _build_path(self, key: str) -> str:
+        """Build the model path for the given key.
+
+        Args:
+            key: the key for which to build the model path.
+
+        Returns:
+            the model path.
+
+        Raises:
+            AssertionError: if the directory, file name, or the given key is
+                not set or allowed.
+        """
+        assert self.directory, "Directory must be set."
+        assert self.file_name, "File name must be set."
+        assert key in self._get_allowed_keys(), f"Key {key} is not allowed."
+        return f"{self.directory}/{key}/{self.file_name}"
+
+
+@dataclass
+class _Model(Generic[T]):
+    """Represent a machine learning model."""
+
+    data: T
+    """The machine learning model."""
+
+    path: str
+    """The path to the model file."""
+
+
+class ModelManager(Generic[T]):
+    """Manage machine learning models.
+
+    The model manager acts as a central hub for retrieving models, reloading
+    models, and monitoring model paths for changes.
+    """
+
+    def __init__(self, model_repository: ModelRepository) -> None:
+        """Initialize the model manager.
+
+        Args:
+            model_repository: a model repository that defines the mapping
+                between model keys and paths.
+        """
+        self._models: dict[Any, _Model[T]] = {}
+
+        models_config = model_repository.get_models_config()
+        for key, model_path in models_config.items():
+            self._models[key] = _Model(data=self._load(model_path), path=model_path)
+
+        self._monitoring_task: asyncio.Task[None] = asyncio.create_task(
+            self._start_model_monitor()
+        )
+
+    @classmethod
+    def single(cls, model_path: str) -> "ModelManager[T]":
```

Comment:
I would not have these classmethods here. In the end they don't save much code.

Reply:
right, it does not add much value with the new `ModelRepository`